### PR TITLE
treewide: markdown option docs

### PR DESCRIPTION
**Pull request by pennae**
###### Description of changes

just an attempt at getting option descriptions to markdown (cf #175586).

the conversion is obviously not completely faithful (since representing the difference between `<filename>` and `<literal>` isn't easy to represent in markdown, let alone very useful?), but minimizing the diff between the current options.xml and the one after md conversions is an explicit goal. once all descriptions have been converted we can drop `mdDoc` again in a mechanical patch.

on top of the syntax extensions the manual already uses this introduces two new ones: `` {command}`foo` `` and `` {option}`foo` ``. both render to their docbook `<tag>foo</tag>` equivalent to keep the conversion between docbook descriptions and md as faithful as possible. in the future we may want to use `{option}` to auto-link option uses to their descriptions, but right now `{command}` and `{option}` have no such special functions.

one problem we see with these extensions is that the processor used for the options manual would have to support them, which at least for mddoc seems very hard to achieve at first glance. keeping `{option}` doesn't seem that important, but `{command}` is rendered differently from regular literals in the manpage build (so we should probably keep that, somehow). keeping the admonitions also seems very useful, not least because they stand out nicely even in the manpage build.

cc @roberth 

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by roberth** [comment]
This fits all the requirements expressed in #175586 at this point, with the exception of @edolstra's dislike of `mdDoc`, although that was in conjunction with pandoc.

I suppose another potential complaint is that we're using a markdown processor, `mistune`, that isn't the one used for the rest of the manual, and isn't the one used in the Nix (package) manual either. It [is](https://github.com/lepture/mistune/blob/master/docs/index.rst) "compatible with sane CommonMark rules."
We may want to review the [plugins](https://github.com/lepture/mistune/blob/master/docs/plugins.rst) but otherwise it seems we're good.
If someone can find a good reason to switch to a different markdown processor, they are free to do so during or after this PR, but that should not block migration progress.

The implementation looks ok so far. Will give this a try later.

Looks like we can finally complete the markdown migration! :tada: 

-----


> the difference between `<filename>` and `<literal>` isn't easy to represent in markdown

tl;dr not a problem.
I am not aware of anything that uses this information. The only potential use I can think of is producing links somehow, but it doesn't really hold enough information to produce such links. If we want links, we should just add links.

File: nixos/lib/make-options-doc/mergeJSON.py [hunk 1]
```diff
@@ -41,6 +41,49 @@ def unpivot(options: Dict[Key, Option]) -> Dict[str, JSON]:
...
+    for (name, option) in options.items():
+        if 'description' not in option:
+            continue
+        if type(option['description']) != dict:
+            continue
+        if '_type' not in option['description']:
+            continue
+        if option['description']['_type'] != 'mdDoc':
+            continue
+        text = option['description']['text']
```
Comment (roberth) [inline nixos/lib/make-options-doc/mergeJSON.py:78]: We should also do `defaultText` and `example` when they have `literalDocBook`.

File: nixos/lib/make-options-doc/mergeJSON.py [hunk 2]
```diff
@@ -41,6 +41,49 @@ def unpivot(options: Dict[Key, Option]) -> Dict[str, JSON]:
...
+        ast = mistune.markdown(text, renderer='ast')
+        # pass the double line break for postprocess-option-descriptions.xsl.
+        # also keep trailing spaces so we can diff the generated XML to check for
+        # conversion bugs.
```
Comment (roberth) [inline nixos/lib/make-options-doc/mergeJSON.py:82]: Diffing seems like a good idea.

Instead of trying to match every quirk, we can choose to canonicalize both inputs to the diff. We'll need something like that for `<filename>` anyway.

File: nixos/lib/make-options-doc/mergeJSON.py [hunk 3]
```diff
@@ -41,6 +41,49 @@ def unpivot(options: Dict[Key, Option]) -> Dict[str, JSON]:
...
+
+    def apply(node) -> str:
+        if node['type'] == 'paragraph':
+            return "".join(map(apply, node['children']))
+        elif node['type'] == 'text':
+            return escape(node['text'])
+        elif node['type'] == 'codespan':
+            return f"<literal>{escape(node['text'])}</literal>"
+        elif node['type'] == 'block_code':
+            return f"<programlisting>{escape(node['text'])}</programlisting>"
```
Comment (jtojnar) [inline nixos/lib/make-options-doc/mergeJSON.py:57]: Would be nice to include the [`language` attribute](https://tdg.docbook.org/tdg/5.1/programlisting.html#element.db.programlisting.5.4).

File: nixos/lib/make-options-doc/mergeJSON.py [hunk 4]
```diff
@@ -41,6 +41,49 @@ def unpivot(options: Dict[Key, Option]) -> Dict[str, JSON]:
...
+        elif node['type'] == 'link':
+            if node['link'][0:1] == '#':
+                attr = "linkend"
+                link = node['link'][1:]
+            else:
+                attr = "xlink:href"
+                link = escape(node['link'])
+            return f"<link {attr}=\"{link}\">{''.join(map(apply, node['children']))}</link>"
```
Comment (jtojnar) [inline nixos/lib/make-options-doc/mergeJSON.py:65]: [`escape`](https://docs.python.org/3/library/xml.sax.utils.html#xml.sax.saxutils.escape) is only for element bodies, not attributes. You would need to use [`quoteattr`](https://docs.python.org/3/library/xml.sax.utils.html#xml.sax.saxutils.quoteattr). Also escape `link` in both branches.

File: nixos/modules/services/networking/syncthing.nix [hunk 5]
```diff
@@ -135,27 +135,27 @@ in {
...
 
             introducer = mkOption {
               type = types.bool;
               default = false;
-              description = ''
+              description = mdDoc ''
                 Whether the device should act as an introducer and be allowed
                 to add folders on this computer.
-                See <link xlink:href="https://docs.syncthing.net/users/introducer.html"/>.
+                See [](https://docs.syncthing.net/users/introducer.html).
```
Comment (jtojnar) [inline nixos/modules/services/networking/syncthing.nix:149]: The [auto link-text extension](https://nixos.org/manual/nixpkgs/stable/#ssec-contributing-markup-automatic-links) only really works for in-document cross-references. The markdown syntax for external links is `<https://docs.syncthing.net/users/introducer.html>` (or just leaving them as is, when unambiguous).

File: nixos/lib/make-options-doc/mergeJSON.py (hunk 1 above)
Comment (pennae) [reply nixos/lib/make-options-doc/mergeJSON.py:78]: yes! this is only a proof of concept to see whether the approach is worth doing. all other docbook tags used by nixos docs and the remaining types of fragments will be added.

File: nixos/modules/services/networking/syncthing.nix (hunk 5 above)
Comment (pennae) [reply nixos/modules/services/networking/syncthing.nix:149]: leaving them as is has not worked in this particular instance, but we'll try the `<>` variant and change it to that if successful. thank you

**Comment by pennae** [comment]
> I suppose another potential complaint is that we're using a markdown processor, `mistune`, that isn't the one used for the rest of the manual, and isn't the one used in the Nix (package) manual either. It [is](https://github.com/lepture/mistune/blob/master/docs/index.rst) "compatible with sane CommonMark rules."
> We may want to review the [plugins](https://github.com/lepture/mistune/blob/master/docs/plugins.rst) but otherwise it seems we're good.

we really only chose mistune because it's a commonmark parser available in python that doesn't enlarge the docs closure by too much. anything else that can parse markdown into an ast would be fine. even shelling out to the favored markdown processor and have it dump an ast wouldn't be a problem, we think.

**Comment by pennae** [comment]
* added `literalMD` for examples and defaults
* fixed link handling and escaping
* keep `language` attrs on `programlisting` (no option uses this so far)
* added lists and emphasis

things we're not sure how to convert:
* `<option>`? probably best to turn them into local links
* `<replaceable>`? would suggest `<replaceable>X</replaceable> → «X»` since this matches how attrset placeholders are rendered
* `<note>`/`<warning>`/etc? nixos markdown has admonitions that would be useful here, maybe mistune can be made to support them with plugins/parser rules. haven't checked that yet

**Comment by pennae** [comment]
admonitions are indeed possible :)

**Comment by pennae** [comment]
updated description to the current state of the patch set. apart from a few kinds of admonitions the conversion script should be mostly complete (if still very experimental).

**Comment by pennae** [comment]
does this need more modules converted as examples? :thinking:

**Comment by roberth** [comment]
- [x] converted entries render as usual
 - [x] non-converted entries render as usual
 - [x] implementation looks good

> does this need more modules converted as examples?

I don't think it helps with review and merge. If you're not confident that others can contribute conversion PRs, you could do some more, but leave them for a follow-up PR.

Some documentation will be helpful:
  - [ ] Nixpkgs [Contributing to this documentation/Syntax](https://nixos.org/manual/nixpkgs/unstable/#sec-contributing-markup)
      - this already explains the existing markdown syntax used in manuals. Is it complete for the purpose of options? Any differences?
  - [ ] NixOS [Writing Documentation](https://nixos.org/manual/nixos/unstable/index.html#sec-writing-documentation)
      - reference the Nixpkgs manual? 
  - [ ] NixOS [Contributing to this manual](https://nixos.org/manual/nixos/unstable/index.html#chap-contributing)
      - would be nice to have some links to the other sections
 
When those are done and it's ready for other to start contributing, I think we're ready for a merge.

File: nixos/lib/make-options-doc/mergeJSON.py [hunk 6]
```diff
@@ -41,6 +41,138 @@ def unpivot(options: Dict[Key, Option]) -> Dict[str, JSON]:
...
+                return super(Renderer, self)._get_method(name)
+            except AttributeError:
+                def not_supported(children, **kwargs):
+                    raise NotImplementedError("md node not supported yet", name, children, **kwargs)
+                return not_supported
+
+        def text(self, text):
+            return escape(text)
+        def paragraph(self, text):
+            return text + "\n\n"
```
Comment (jtojnar) [inline nixos/lib/make-options-doc/mergeJSON.py:65]: Should not this be wrapped in `<para>`, example: https://nixos.org/manual/nixos/stable/options.html#opt-services.matrix-synapse.settings

Comment (pennae) [reply nixos/lib/make-options-doc/mergeJSON.py:65]: the `<para>` is generated by the container, eg in line 88. the way this script uses `paragraph` is more akin to paragraph *breaks*, because that best mimics how the current system treats the double newline as a paragraph breaks.

Comment (jtojnar) [reply nixos/lib/make-options-doc/mergeJSON.py:65]: That would not work for the outermost paragraphs though.

----

On a different note, I wonder if we should just do preprocessing here and skip https://github.com/NixOS/nixpkgs/blob/e0e9109c12eaee527fbae2e7ac3d6b79588b26d3/nixos/lib/make-options-doc/postprocess-option-descriptions.xsl. I can’t imagine that to be very fast.

File: nixos/lib/make-options-doc/mergeJSON.py [hunk 7]
```diff
@@ -41,6 +41,138 @@ def unpivot(options: Dict[Key, Option]) -> Dict[str, JSON]:
...
+            return f"<emphasis>{text}</emphasis>"
+        def strong(self, text):
+            return f"<emphasis role=\"strong\">{text}</emphasis>"
+        def admonition(self, text, kind):
+            if kind not in admonitions:
+                raise NotImplementedError(f"admonition {kind} not supported yet")
+            tag = admonitions[kind]
+            # we don't keep whitespace here because usually we'll contain only
+            # a single paragraph and the original docbook string is no longer
+            # available to restore the trailer.
```
Comment (jtojnar) [inline nixos/lib/make-options-doc/mergeJSON.py:101]: What do you mean by trailer?

Comment (pennae) [reply nixos/lib/make-options-doc/mergeJSON.py:101]: trailing whitespace that is not rendered. we keep trailing whitespace at the end of a description to make diffing `options.xml` before and after a migration easier, but that isn't possible here. apart from that there is not reason to care about whitespace and if we use a different diffing process we can also remove this whitespace handling entirely.

File: nixos/lib/make-options-doc/mergeJSON.py (hunk 6 above)
Comment (pennae) [reply nixos/lib/make-options-doc/mergeJSON.py:65]: not sure we understand what you mean. there aren't any "outer" paragraphs as such, the docbook `description` is pasted into a `<para>` element in https://github.com/NixOS/nixpkgs/blob/e0e9109c12eaee527fbae2e7ac3d6b79588b26d3/nixos/lib/make-options-doc/options-to-docbook.xsl#L55. the postprocessing you linked artifically splits that wrapper parapgraph at markdown paragraph boundaries by turning the description into a technically-invalid xml fragment.

as for skipping the preprocessing: don't think that's feasible unless we convert *all* options that use markdown paragraph separation at the same time as well.

**Comment by pennae** [comment]
we've added some documentation to the nixos and nixpkgs manuals clarifying which extensions are available in options docs. not entirely sure which sections to link for the final nixos manual chapter though :confused:

File: nixos/lib/make-options-doc/mergeJSON.py (hunk 6 above)
Comment (jtojnar) [reply nixos/lib/make-options-doc/mergeJSON.py:65]: I meant skipping preprocessing for the markdown content hoping for better performance as less and less DocBook descriptions exist. But it appears the performance impact is negligible anyway.

**Comment by roberth** [comment]
> which sections to link for the final nixos manual chapter though

I was also thinking about documenting the transition process, but it's perhaps better to write an issue for that, similar to the other tracking issues we have, like https://github.com/NixOS/nixpkgs/issues/24288.

**Comment by roberth** [comment]
I've tried it and it worked well.

My only findings: if you get invalid xml errors in "`intermediate.xml`", you've forgot `mdDoc`.

**Comment by pennae** [comment]
> I've tried it and it worked well.

:champagne:!

> My only findings: if you get invalid xml errors in "`intermediate.xml`", you've forgot `mdDoc`.

not sure we can do anything about that, unfortunately. though what we probably could do is print a warning (error?) if a level of options is not exclusive `mdDoc` or exclusively docbook. not sure how useful that would be in the face of option merging though

**Comment by roberth** [comment]
> if a level of options is not exclusive `mdDoc` or exclusively docbook

I'm not sure if it's worth the effort. Option merging is quite rare in NixOS, so I wouldn't expect much trouble from false positives if the error message mentions the possibility. I think it's ok to mention the "bad" error message in the tracking issue. It's a temporary state of affairs anyway, during the transition.

File: nixos/doc/manual/development/option-declarations.section.md [hunk 8]
```diff
@@ -56,7 +56,14 @@ The function `mkOption` accepts the following arguments.
 `description`
 
 :   A textual description of the option, in DocBook format, that will be
-    included in the NixOS manual.
+    included in the NixOS manual. During the migration process from DocBook
+    to CommonMark the description may also be written in CommonMark, but has
+    to be wrapped in `lib.mdDoc` to differentiate it from DocBook. See
+    the nixpkgs manual for [the list of CommonMark extensionds](
```
Comment (ncfavier) [inline nixos/doc/manual/development/option-declarations.section.md:62]:
```suggestion
    the nixpkgs manual for [the list of CommonMark extensions](
```

File: nixos/doc/manual/development/option-declarations.section.md [hunk 9]
```diff
@@ -56,7 +56,14 @@ The function `mkOption` accepts the following arguments.
...
+    https://nixos.org/nixpkgs/manual/#sec-contributing-markup)
+    supported by NixOS documentation.
+
+    New documentation should preferrably be written as CommonMark.
```
Comment (ncfavier) [inline nixos/doc/manual/development/option-declarations.section.md:66]:
```suggestion
    New documentation should preferably be written as CommonMark.
```

**Comment by pennae** [comment]
invalid xml errors are probably most often caused by `<⋯>` links, those are easy to spot. the rest of the MD syntax should usually be valid xml, but hopefully we'll catch that in review. and as you say, temporart state—worst case, it's slightly less sightly until we complete the migration. (which hopefully happens before 22.11!)

File: doc/build-aux/pandoc-filters/docbook-writer/rst-roles.lua [hunk 10]
```diff
@@ -32,5 +32,9 @@ function Code(elem)
     if tag ~= nil then
       return pandoc.RawInline('docbook', '<' .. tag .. '>' .. content .. '</' .. tag .. '>')
     end
+  elseif elem.attributes['role'] == 'command' then
+    return pandoc.RawInline('docbook', '<command>' .. content .. '</command>')
+  elseif elem.attributes['role'] == 'option' then
+    return pandoc.RawInline('docbook', '<option>' .. content .. '</option>')
```
Comment (pennae) [inline doc/build-aux/pandoc-filters/docbook-writer/rst-roles.lua:38]: these should move upward, next to the block that handles `{file}`

**Review by roberth — APPROVED** [review]

**Review by ncfavier — APPROVED** [review]

**Comment by peterhoeg** [comment]
For reasons unknown (to me), I'm getting this:
```
error: builder for '/nix/store/ajymzc86n3gqanmcj2kiwqra49r825g5-lazy-options.json.drv' failed with exit code 1;                                                                                                  last 10 log lines:                                                                                                                                                                                        >                                                                                                                                                                                                  
       >           252|         type = nullOr str; 
       >           253|         description = mdDoc ''
       >              |         ^
       >           254|           Address to listen on. Listen on `0.0.0.0`/`::`
       > Cacheable portion of option doc build failed.   
       > Usually this means that an option attribute that ends up in documentation (eg `default` or `description`) depends on the restricted module arguments `config` or `pkgs`.
```

If I drop all the `mdDoc` references from `nixos/modules/services/networking/mosquitto.nix`, I can use `nixos-rebuild` again.

**Comment by pennae** [comment]
are you backporting newer modules to older revisions? does it work with `documentation.nixos.options.splitBuild = false`?

**Comment by roberth** [comment]
@pennae, found some improvements #179351
(not related to the previous two messages)

**Comment by peterhoeg** [comment]
> are you backporting newer modules to older revisions? 

This is admittedly with a custom nixpkgs, but it's building the latest unstable. The problem starting showing about a week ago.

> does it work with documentation.nixos.options.splitBuild = false?

I'll try to create the smallest possible example that reproduces it and let you guys know. In the meantime I'm just patching out `mdDoc`.
